File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int d = 0;
    public int diameterOfBinaryTree(TreeNode root) {
        rec(root);
        return d;
    }
    private int rec(TreeNode node){
        if(node == null) return 0;

        int left = rec(node.left);
        int right = rec(node.right);

        d = Math.max(d, left + right);

        return Math.max(left, right) + 1;
    }
}

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

var d int

func diameterOfBinaryTree(root *TreeNode) int {
    d = 0
    rec(root)
    return d
}

func rec(node *TreeNode) int {
    if node == nil { return 0 }

    left := rec(node.Left)
    right := rec(node.Right)

    d = Max(d, left+right)

    return Max(left, right) + 1
}

func Max(a, b int) int {
    if a > b { return a }
    return b
}